# Traditional Japanese Restaurant in USA

## Introduction

My best friend "Aya", the young Japanese lady just graduated from university have plan to marry with her partner 
and will be moved to New York City after married because her partner working there as a business man. 
Aya has her dream to open up her own Japanese restaurant and this is a chance to open her own restaurant 
in the amazimg place New York City!

## Business Problem

Aya doesn't have any idea because she doesn't know anything about New York. 
The important thing she have to do is exploring about New York before doing her Business Plan.

## Describe the data that will be using to execute the idea.

For this project I plan to use the Foursquare location data to execute my idea for Aya to suggest her where is the best area of New York city she should open her Japanese restaurant.

Based on definition of my idea, factors that will influence my decission are:

- number of existing restaurants in the neighborhood (any type of restaurant)
- number of and distance to Japanese restaurants in the neighborhood, if any
- distance of neighborhood from city center I decided to use regularly spaced grid of locations, centered around city center, to define our neighborhoods.

Following data sources will be needed to extract/generate the required information:

- Centers of candidate areas will be generated algorithmically and approximate addresses of centers of those areas will be obtained using New York Data that already used by previous lab.
- Number of restaurants and their type and location in every neighborhood will be obtained using Foursquare API

## Neighborhood Candidates

Create latitude & longitude coordinates for centroids of candidate neighborhoods. 
- first, I will explore the New York City
- then explore Manhattan, the area that Aya's partner house is located in.

### Before start the project, I will import the necessary libraries

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes 
import folium # map rendering library
import requests # library to handle requests
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner
import random # library for random number generation

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 
    
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize

!conda install -c conda-forge folium=0.5.0 --yes
import folium # plotting library

!pip install shapely
import shapely.geometry

!pip install pyproj
import pyproj

import math

print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.

Libraries imported.


#### Download and explore New York dataset

In [2]:
!wget -q -O 'newyork_data.json' https://cocl.us/new_york_dataset
print('Data downloaded!')

Data downloaded!


Next, load the data

In [3]:
with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)

In [4]:
newyork_data

{'type': 'FeatureCollection',
 'totalFeatures': 306,
 'features': [{'type': 'Feature',
   'id': 'nyu_2451_34572.1',
   'geometry': {'type': 'Point',
    'coordinates': [-73.84720052054902, 40.89470517661]},
   'geometry_name': 'geom',
   'properties': {'name': 'Wakefield',
    'stacked': 1,
    'annoline1': 'Wakefield',
    'annoline2': None,
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.84720052054902,
     40.89470517661,
     -73.84720052054902,
     40.89470517661]}},
  {'type': 'Feature',
   'id': 'nyu_2451_34572.2',
   'geometry': {'type': 'Point',
    'coordinates': [-73.82993910812398, 40.87429419303012]},
   'geometry_name': 'geom',
   'properties': {'name': 'Co-op City',
    'stacked': 2,
    'annoline1': 'Co-op',
    'annoline2': 'City',
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.82993910812398,
     40.87429419303012,
     -73.82993910812398,
     40.87429419303012]}},
  {'type': 'Feature',
 

Notice how all the relevant data is in the features key, which is basically a list of the neighborhoods. I will define a new variable that includes this data.

In [5]:
neighborhoods_data = newyork_data['features']

Take a look at the first item in this list.

In [6]:
neighborhoods_data[0]

{'type': 'Feature',
 'id': 'nyu_2451_34572.1',
 'geometry': {'type': 'Point',
  'coordinates': [-73.84720052054902, 40.89470517661]},
 'geometry_name': 'geom',
 'properties': {'name': 'Wakefield',
  'stacked': 1,
  'annoline1': 'Wakefield',
  'annoline2': None,
  'annoline3': None,
  'annoangle': 0.0,
  'borough': 'Bronx',
  'bbox': [-73.84720052054902,
   40.89470517661,
   -73.84720052054902,
   40.89470517661]}}

Then , tranform the data into a pandas dataframe

In [7]:
# define the dataframe columns
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

# instantiate the dataframe
neighborhoods = pd.DataFrame(columns=column_names)

In [8]:
neighborhoods

Empty DataFrame
Columns: [Borough, Neighborhood, Latitude, Longitude]
Index: []

Loop through the data and fill the dataframe one row at a time

In [10]:
for data in neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    neighborhoods = neighborhoods.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)

In [11]:
neighborhoods.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

Make sure that the dataset has all 5 boroughs and 306 neighborhoods.

In [13]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)

The dataframe has 5 boroughs and 306 neighborhoods.


Use geopy library to get the latitude and longitude values of New York City.

In [14]:
address = 'New York City, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of New York City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of New York City are 40.7127281, -74.0060152.


Create a map of New York with neighborhoods superimposed on top.

In [15]:
# create map of New York using latitude and longitude values
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

However, Aya's partner's home is located in Manhattan so I will scope only in Manhattan

In [17]:
manhattan_data = neighborhoods[neighborhoods['Borough'] == 'Manhattan'].reset_index(drop=True)
manhattan_data.head()

Borough        Neighborhood   Latitude  Longitude
0  Manhattan         Marble Hill  40.876551 -73.910660
1  Manhattan           Chinatown  40.715618 -73.994279
2  Manhattan  Washington Heights  40.851903 -73.936900
3  Manhattan              Inwood  40.867684 -73.921210
4  Manhattan    Hamilton Heights  40.823604 -73.949688

Get the geographical coordinates of Manhattan

In [18]:
address = 'Manhattan, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Manhattan are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Manhattan are 40.7896239, -73.9598939.


As I did with all of New York City, next is visualize Manhattan the neighborhoods in it.

In [19]:
# create map of Manhattan using latitude and longitude values
map_manhattan = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(manhattan_data['Latitude'], manhattan_data['Longitude'], manhattan_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_manhattan)  
    
map_manhattan

Next, I will going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

#### Define Foursquare Credentials and Version

In [20]:
CLIENT_ID = 'ICB5PI2AB2TX22SZECS1WOLPX3EUWF3LSLGWONXVE3BDPEMS' # your Foursquare ID
CLIENT_SECRET = 'LPPUSMAN43VJP3PYQ5TIXAK5EGUILNQCKUSMUJSNENB2FKYO' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: ICB5PI2AB2TX22SZECS1WOLPX3EUWF3LSLGWONXVE3BDPEMS
CLIENT_SECRET:LPPUSMAN43VJP3PYQ5TIXAK5EGUILNQCKUSMUJSNENB2FKYO


Let's explore the first neighborhood in our dataframe.

Get the neighborhood's name.

In [21]:
manhattan_data.loc[0, 'Neighborhood']

'Marble Hill'

Get the neighborhood's latitude and longitude values.

In [22]:
neighborhood_latitude = manhattan_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = manhattan_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = manhattan_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Marble Hill are 40.87655077879964, -73.91065965862981.


#### Now, let's get the top 500 venues that are in Marble Hill within a radius of 20 KM.

In [33]:
# type your answer here
LIMIT = 500 # limit of number of venues returned by Foursquare API
radius = 20000 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=ICB5PI2AB2TX22SZECS1WOLPX3EUWF3LSLGWONXVE3BDPEMS&client_secret=LPPUSMAN43VJP3PYQ5TIXAK5EGUILNQCKUSMUJSNENB2FKYO&v=20180605&ll=40.87655077879964,-73.91065965862981&radius=20000&limit=500'

Send the GET request and examine the resutls

In [34]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5da58c6f5d891b002c800b50'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'},
    {'name': '$-$$$$', 'key': 'price'}]},
  'headerLocation': 'New York',
  'headerFullLocation': 'New York',
  'headerLocationGranularity': 'city',
  'totalResults': 240,
  'suggestedBounds': {'ne': {'lat': 41.05655095879982,
    'lng': -73.67304688055954},
   'sw': {'lat': 40.69655059879946, 'lng': -74.14827243670008}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '58935fd798f8aa7c14662653',
       'name': 'Kingsbridge Social Club',
       'location': {'address': '3625 Kingsbridge Ave',
        'lat': 40.88454548045115,
        'lng': -73.90196378126393,
        'labeledLatLngs': [{'label'

Borrow the get_category_type function

In [35]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Clean the json and structure it into a pandas dataframe.

In [36]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name      categories        lat        lng
0  Kingsbridge Social Club     Pizza Place  40.884545 -73.901964
1         Inwood Hill Park            Park  40.870858 -73.920129
2                  ChocNYC          Bakery  40.868396 -73.918742
3    Inwood Farmers Market  Farmers Market  40.869062 -73.920560
4           Bronx Alehouse        Beer Bar  40.884749 -73.899699

Check how many venues were returned by Foursquare?

In [37]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

100 venues were returned by Foursquare.


#### Explore Neighborhoods in Manhattan

Create a function to repeat the same process to all the neighborhoods in Manhattan

In [38]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Write the code to run the above function on each neighborhood and create a new dataframe called manhattan_venues.

In [40]:
manhattan_venues = getNearbyVenues(names=manhattan_data['Neighborhood'],
                                   latitudes=manhattan_data['Latitude'],
                                   longitudes=manhattan_data['Longitude']
                                  )


Marble Hill
Chinatown
Washington Heights
Inwood
Hamilton Heights
Manhattanville
Central Harlem
East Harlem
Upper East Side
Yorkville
Lenox Hill
Roosevelt Island
Upper West Side
Lincoln Square
Clinton
Midtown
Murray Hill
Chelsea
Greenwich Village
East Village
Lower East Side
Tribeca
Little Italy
Soho
West Village
Manhattan Valley
Morningside Heights
Gramercy
Battery Park City
Financial District
Carnegie Hill
Noho
Civic Center
Midtown South
Sutton Place
Turtle Bay
Tudor City
Stuyvesant Town
Flatiron
Hudson Yards


Check the size of the resulting dataframe

In [41]:
print(manhattan_venues.shape)
manhattan_venues.head()

(3321, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude          Venue  \
0  Marble Hill              40.876551               -73.91066       Arturo's   
1  Marble Hill              40.876551               -73.91066    Bikram Yoga   
2  Marble Hill              40.876551               -73.91066  Tibbett Diner   
3  Marble Hill              40.876551               -73.91066      Starbucks   
4  Marble Hill              40.876551               -73.91066        Dunkin'   

   Venue Latitude  Venue Longitude Venue Category  
0       40.874412       -73.910271    Pizza Place  
1       40.876844       -73.906204    Yoga Studio  
2       40.880404       -73.908937          Diner  
3       40.877531       -73.905582    Coffee Shop  
4       40.877136       -73.906666     Donut Shop

Check how many venues were returned for each neighborhood

In [42]:
manhattan_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
Neighborhood                                                                
Battery Park City                       99                      99     99   
Carnegie Hill                          100                     100    100   
Central Harlem                          46                      46     46   
Chelsea                                100                     100    100   
Chinatown                              100                     100    100   
Civic Center                           100                     100    100   
Clinton                                100                     100    100   
East Harlem                             42                      42     42   
East Village                           100                     100    100   
Financial District                     100                     100    100   
Flatiron                               100                     100    100   
Gramercy                               100                     100    100   
Greenwich Village                      100                     100    100   
Hamilton Heights                        60                      60     60   
Hudson Yards                            82                      82     82   
Inwood                                  58                      58     58   
Lenox Hill                             100                     100    100   
Lincoln Square                         100                     100    100   
Little Italy                           100                     100    100   
Lower East Side                         55                      55     55   
Manhattan Valley                        60                      60     60   
Manhattanville                          41                      41     41   
Marble Hill                             23                      23     23   
Midtown                                100                     100    100   
Midtown South                          100                     100    100   
Morningside Heights                     41                      41     41   
Murray Hill                            100                     100    100   
Noho                                   100                     100    100   
Roosevelt Island                        26                      26     26   
Soho                                   100                     100    100   
Stuyvesant Town                         22                      22     22   
Sutton Place                           100                     100    100   
Tribeca                                100                     100    100   
Tudor City                              80                      80     80   
Turtle Bay                             100                     100    100   
Upper East Side                        100                     100    100   
Upper West Side                        100                     100    100   
Washington Heights                      86                      86     86   
West Village                           100                     100    100   
Yorkville                              100                     100    100   

                     Venue Latitude  Venue Longitude  Venue Category  
Neighborhood                                                          
Battery Park City                99               99              99  
Carnegie Hill                   100              100             100  
Central Harlem                   46               46              46  
Chelsea                         100              100             100  
Chinatown                       100              100             100  
Civic Center                    100              100             100  
Clinton                         100              100             100  
East Harlem                      42               42              42  
East Village                    100              100             100  
Finan

Find out how many unique categories can be curated from all the returned venues

In [43]:
print('There are {} uniques categories.'.format(len(manhattan_venues['Venue Category'].unique())))

There are 334 uniques categories.


### Analyze Each Neighborhood

In [44]:
# one hot encoding
manhattan_onehot = pd.get_dummies(manhattan_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
manhattan_onehot['Neighborhood'] = manhattan_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [manhattan_onehot.columns[-1]] + list(manhattan_onehot.columns[:-1])
manhattan_onehot = manhattan_onehot[fixed_columns]

manhattan_onehot.head()

Neighborhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0  Marble Hill                  0               0                  0   
1  Marble Hill                  0               0                  0   
2  Marble Hill                  0               0                  0   
3  Marble Hill                  0               0                  0   
4  Marble Hill                  0               0                  0   

   African Restaurant  American Restaurant  Antique Shop  Arcade  \
0                   0                    0             0       0   
1                   0                    0             0       0   
2                   0                    0             0       0   
3                   0                    0             0       0   
4                   0                    0             0       0   

   Arepa Restaurant  Argentinian Restaurant  Art Gallery  Art Museum  \
0                 0                       0            0           0   
1                 0                       0            0           0   
2                 0                       0            0           0   
3                 0                       0            0           0   
4                 0                       0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auditorium  \
0                    0                 0                   0           0   
1                    0                 0                   0           0   
2                    0                 0                   0           0   
3                    0                 0                   0           0   
4                    0                 0                   0           0   

   Australian Restaurant  Austrian Restaurant  Auto Workshop  BBQ Joint  \
0                      0                    0              0          0   
1                      0                    0              0          0   
2                      0                    0              0          0   
3                      0                    0              0          0   
4                      0                    0              0          0   

   Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0           0           0       0     0    0               0   
1           0           0       0     0    0               0   
2           0           0       0     0    0               0   
3           0           0       0     0    0               0   
4           0           0       0     0    0               0   

   Basketball Court  Bed & Breakfast  Beer Bar  Beer Garden  Beer Store  \
0                 0                0         0            0           0   
1                 0                0         0            0           0   
2                 0                0         0            0           0   
3                 0                0         0            0           0   
4                 0                0         0            0           0   

   Big Box Store  Bike Rental / Bike Share  Bike Shop  Bike Trail  Bistro  \
0              0                         0          0           0       0   
1              0                         0          0           0       0   
2              0                         0          0           0       0   
3              0                         0          0           0       0   
4              0                         0          0           0       0   

   Board Shop  Boat or Ferry  Bookstore  Boutique  Boxing Gym  \
0           0              0          0         0           0   
1           0              0          0         0           0   
2           0              0          0         0           0   
3           0              0          0         0           0   
4           0              0          0         0           0   

   Brazilian Restaurant  Breakfast Spot  Bridal Shop  Bridge  Bubble Tea Shop  \
0                     0               0         

Examine the new dataframe size

In [45]:
manhattan_onehot.shape

(3321, 335)

### Next, group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [46]:
manhattan_grouped = manhattan_onehot.groupby('Neighborhood').mean().reset_index()
manhattan_grouped

Neighborhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0     Battery Park City           0.000000            0.00               0.00   
1         Carnegie Hill           0.000000            0.00               0.00   
2        Central Harlem           0.000000            0.00               0.00   
3               Chelsea           0.000000            0.00               0.00   
4             Chinatown           0.000000            0.00               0.00   
5          Civic Center           0.000000            0.00               0.00   
6               Clinton           0.000000            0.00               0.00   
7           East Harlem           0.000000            0.00               0.00   
8          East Village           0.000000            0.00               0.00   
9    Financial District           0.010000            0.00               0.00   
10             Flatiron           0.000000            0.00               0.00   
11             Gramercy           0.000000            0.00               0.00   
12    Greenwich Village           0.000000            0.00               0.00   
13     Hamilton Heights           0.000000            0.00               0.00   
14         Hudson Yards           0.000000            0.00               0.00   
15               Inwood           0.000000            0.00               0.00   
16           Lenox Hill           0.000000            0.00               0.01   
17       Lincoln Square           0.000000            0.00               0.00   
18         Little Italy           0.000000            0.00               0.00   
19      Lower East Side           0.000000            0.00               0.00   
20     Manhattan Valley           0.000000            0.00               0.00   
21       Manhattanville           0.000000            0.00               0.00   
22          Marble Hill           0.000000            0.00               0.00   
23              Midtown           0.000000            0.00               0.00   
24        Midtown South           0.000000            0.00               0.00   
25  Morningside Heights           0.000000            0.00               0.00   
26          Murray Hill           0.000000            0.00               0.00   
27                 Noho           0.000000            0.00               0.00   
28     Roosevelt Island           0.000000            0.00               0.00   
29                 Soho           0.010000            0.00               0.00   
30      Stuyvesant Town           0.000000            0.00               0.00   
31         Sutton Place           0.000000            0.01               0.00   
32              Tribeca           0.000000            0.00               0.00   
33           Tudor City           0.000000            0.00               0.00   
34           Turtle Bay           0.000000            0.00               0.00   
35      Upper East Side           0.000000            0.00               0.00   
36      Upper West Side           0.010000            0.00               0.00   
37   Washington Heights           0.011628            0.00               0.00   
38         West Village           0.010000            0.00               0.00   
39            Yorkville           0.000000            0.00               0.00   

    African Restaurant  American Restaurant  Antique Shop  Arcade  \
0             0.000000             0.010101          0.00    0.00   
1             0.000000             0.010000          0.00    0.00   
2             0.043478             0.043478          0.00    0.00   
3             0.000000             0.030000          0.00    0.00   
4             0.000000             0.040000          0.00    0.00   
5             0.000000             0.030000          0.01    0.00   
6             0.000000             0.040000          0.00    0.00   
7             0.000000             0.000000          0.00    0.00   
8             0.000000             0.020000          0.01    0.00   

Confirm the new size

In [47]:
manhattan_grouped.shape

(40, 335)

Print each neighborhood along with the top 10 most common venues

In [48]:
num_top_venues = 10

for hood in manhattan_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = manhattan_grouped[manhattan_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Battery Park City----
            venue  freq
0            Park  0.08
1     Coffee Shop  0.07
2           Hotel  0.05
3             Gym  0.04
4   Memorial Site  0.04
5   Boat or Ferry  0.03
6       Wine Shop  0.03
7      Playground  0.02
8   Grocery Store  0.02
9  Ice Cream Shop  0.02


----Carnegie Hill----
               venue  freq
0        Coffee Shop  0.06
1        Pizza Place  0.06
2               Café  0.05
3                Gym  0.03
4                Spa  0.03
5          Bookstore  0.03
6     Cosmetics Shop  0.03
7             Bakery  0.03
8  French Restaurant  0.03
9      Grocery Store  0.03


----Central Harlem----
                 venue  freq
0   Seafood Restaurant  0.04
1          Art Gallery  0.04
2    French Restaurant  0.04
3       Cosmetics Shop  0.04
4                  Bar  0.04
5   Chinese Restaurant  0.04
6  American Restaurant  0.04
7   African Restaurant  0.04
8                  Gym  0.02
9            Cafeteria  0.02


----Chelsea----
                 venue  fre

Put into a pandas dataframe

First, write a function to sort the venues in descending order.

In [49]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Create the new dataframe and display the top 5 venues for each neighborhood.

In [61]:
num_top_venues = 5

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = manhattan_grouped['Neighborhood']

for ind in np.arange(manhattan_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(manhattan_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0  Battery Park City                  Park           Coffee Shop   
1      Carnegie Hill           Coffee Shop           Pizza Place   
2     Central Harlem    Chinese Restaurant    African Restaurant   
3            Chelsea           Coffee Shop        Ice Cream Shop   
4          Chinatown    Chinese Restaurant          Cocktail Bar   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  
0                 Hotel         Memorial Site                   Gym  
1                  Café           Yoga Studio   Japanese Restaurant  
2   American Restaurant                   Bar        Cosmetics Shop  
3                Bakery    Italian Restaurant             Nightclub  
4   American Restaurant    Salon / Barbershop                Bakery

#### Cluster Neighborhoods

Run k-means to cluster the neighborhood into 3 clusters.

In [62]:
# set number of clusters
kclusters = 3

manhattan_grouped_clustering = manhattan_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(manhattan_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 1, 0, 0, 0, 0, 0, 1, 1, 0], dtype=int32)

Create a new dataframe that includes the cluster as well as the top 5 venues for each neighborhood.

In [63]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

manhattan_merged = manhattan_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
manhattan_merged = manhattan_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

manhattan_merged.head() # check the last columns!

Borough        Neighborhood   Latitude  Longitude  Cluster Labels  \
0  Manhattan         Marble Hill  40.876551 -73.910660               1   
1  Manhattan           Chinatown  40.715618 -73.994279               0   
2  Manhattan  Washington Heights  40.851903 -73.936900               1   
3  Manhattan              Inwood  40.867684 -73.921210               1   
4  Manhattan    Hamilton Heights  40.823604 -73.949688               1   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0        Sandwich Place           Coffee Shop           Yoga Studio   
1    Chinese Restaurant          Cocktail Bar   American Restaurant   
2     Mobile Phone Shop                Bakery                  Café   
3    Mexican Restaurant                  Café           Pizza Place   
4           Pizza Place                  Café    Mexican Restaurant   

  4th Most Common Venue 5th Most Common Venue  
0         Deli / Bodega       Supplement Shop  
1    Salon / Barbershop                Bakery  
2         Grocery Store                  Bank  
3                Lounge              Pharmacy  
4         Deli / Bodega           Coffee Shop

Finally, visualize the resulting clusters

In [64]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(manhattan_merged['Latitude'], manhattan_merged['Longitude'], manhattan_merged['Neighborhood'], manhattan_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=10,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### Examine Clusters

Examine each cluster and determine the discriminating venue categories that distinguish each cluster. 

### Cluster 1

In [65]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 0, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue    2nd Most Common Venue  \
1             Chinatown    Chinese Restaurant             Cocktail Bar   
5        Manhattanville           Coffee Shop       Italian Restaurant   
6        Central Harlem    Chinese Restaurant       African Restaurant   
8       Upper East Side    Italian Restaurant                  Exhibit   
10           Lenox Hill    Italian Restaurant              Coffee Shop   
11     Roosevelt Island                  Park           Sandwich Place   
12      Upper West Side    Italian Restaurant                      Bar   
13       Lincoln Square               Theater                     Café   
14              Clinton               Theater     Gym / Fitness Center   
15              Midtown                 Hotel      Sporting Goods Shop   
16          Murray Hill           Coffee Shop           Sandwich Place   
17              Chelsea           Coffee Shop           Ice Cream Shop   
18    Greenwich Village    Italian Restaurant         Sushi Restaurant   
21              Tribeca                  Park       Italian Restaurant   
22         Little Italy    Italian Restaurant                   Bakery   
23                 Soho        Clothing Store                 Boutique   
24         West Village    Italian Restaurant  New American Restaurant   
26  Morningside Heights             Bookstore      American Restaurant   
28    Battery Park City                  Park              Coffee Shop   
29   Financial District           Coffee Shop                Wine Shop   
31                 Noho    Italian Restaurant             Cocktail Bar   
32         Civic Center  Gym / Fitness Center        French Restaurant   
33        Midtown South     Korean Restaurant                    Hotel   
34         Sutton Place  Gym / Fitness Center        Indian Restaurant   
35           Turtle Bay    Italian Restaurant                 Wine Bar   
38             Flatiron  Gym / Fitness Center              Yoga Studio   
39         Hudson Yards   American Restaurant       Italian Restaurant   

   3rd Most Common Venue   4th Most Common Venue   5th Most Common Venue  
1    American Restaurant      Salon / Barbershop                  Bakery  
5     Seafood Restaurant      Mexican Restaurant                    Park  
6    American Restaurant                     Bar          Cosmetics Shop  
8                 Bakery             Art Gallery             Coffee Shop  
10           Pizza Place        Sushi Restaurant            Burger Joint  
11           Coffee Shop   Outdoors & Recreation                  School  
12              Wine Bar             Coffee Shop          Cosmetics Shop  
13  Gym / Fitness Center            Concert Hall      Italian Restaurant  
14   American Restaurant      Italian Restaurant               Wine Shop  
15               Theater          Clothing Store             Coffee Shop  
16   Japanese Restaurant                   Hotel      Italian Restaurant  
17                Bakery      Italian Restaurant               Nightclub  
18        Clothing Store       Indian Restaurant       French Restaurant  
21                  Café                     Spa     American Restaurant  
22                  Café      Salon / Barbershop          Sandwich Place  
23           Art Gallery              Shoe Store  Furniture / Home Store  
24        Cosmetics Shop                    Park     American Restaurant  
26           Coffee Shop                    Park            Burger Joint  
28                 Hotel           Memorial Site                     Gym  
29                 Hotel     American Restaurant             Pizza Place  
31           Art Gallery       French Restaurant      Mexican Restaurant  
32                 Hotel             Coffee Shop      Italian Restaurant  
33             Hotel Bar     Japanese Restaurant          Cosmetics Shop  
34    Italian Restaurant  Furniture / Home Store     American Restaurant  
35            Steakhouse        Sushi Restaurant        Ramen R

### Cluster 2

In [66]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 1, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0          Marble Hill        Sandwich Place           Coffee Shop   
2   Washington Heights     Mobile Phone Shop                Bakery   
3               Inwood    Mexican Restaurant                  Café   
4     Hamilton Heights           Pizza Place                  Café   
7          East Harlem    Mexican Restaurant         Deli / Bodega   
9            Yorkville    Italian Restaurant                   Gym   
19        East Village                   Bar              Wine Bar   
20     Lower East Side           Pizza Place                  Café   
25    Manhattan Valley           Coffee Shop                   Bar   
27            Gramercy                   Bar    Italian Restaurant   
30       Carnegie Hill           Coffee Shop           Pizza Place   
36          Tudor City                  Café                  Park   

   3rd Most Common Venue      4th Most Common Venue 5th Most Common Venue  
0            Yoga Studio              Deli / Bodega       Supplement Shop  
2                   Café              Grocery Store                  Bank  
3            Pizza Place                     Lounge              Pharmacy  
4     Mexican Restaurant              Deli / Bodega           Coffee Shop  
7                 Bakery  Latin American Restaurant       Thai Restaurant  
9            Coffee Shop                Pizza Place                   Bar  
19        Ice Cream Shop         Chinese Restaurant          Cocktail Bar  
20          Cocktail Bar         Chinese Restaurant   Japanese Restaurant  
25     Indian Restaurant                Pizza Place         Deli / Bodega  
27   American Restaurant                 Bagel Shop           Pizza Place  
30                  Café                Yoga Studio   Japanese Restaurant  
36    Mexican Restaurant           Greek Restaurant         Deli / Bodega

### Cluster 3

In [67]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 2, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
37  Stuyvesant Town         Boat or Ferry                  Park   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  
37                   Bar              Heliport          Tennis Court

## Results and Discussion

This analysis shows about 3 groups of area around Manhattan, New York, USA.
Each cluster groups show the top 5 common venue of each area to bring the idea which area Aya should open her Japanese restaurant.

Result of all this is 40 area containing largest number of potential new restaurant locations based on number of ranking of each area.
Cluster 1 is the popular location and contain the large rating of restaurant-categories.
Cluster 2 is also contain rating of restaurant-categories, but mostly are light meal or beverages.
Cluster 3 is no contain any restaurant type.

If I recommend Aya, I will suggest the area in Cluster 1 to open her restaurant.

## Conclusion

In [73]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 0, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]].head(30)

Neighborhood 1st Most Common Venue    2nd Most Common Venue  \
1             Chinatown    Chinese Restaurant             Cocktail Bar   
5        Manhattanville           Coffee Shop       Italian Restaurant   
6        Central Harlem    Chinese Restaurant       African Restaurant   
8       Upper East Side    Italian Restaurant                  Exhibit   
10           Lenox Hill    Italian Restaurant              Coffee Shop   
11     Roosevelt Island                  Park           Sandwich Place   
12      Upper West Side    Italian Restaurant                      Bar   
13       Lincoln Square               Theater                     Café   
14              Clinton               Theater     Gym / Fitness Center   
15              Midtown                 Hotel      Sporting Goods Shop   
16          Murray Hill           Coffee Shop           Sandwich Place   
17              Chelsea           Coffee Shop           Ice Cream Shop   
18    Greenwich Village    Italian Restaurant         Sushi Restaurant   
21              Tribeca                  Park       Italian Restaurant   
22         Little Italy    Italian Restaurant                   Bakery   
23                 Soho        Clothing Store                 Boutique   
24         West Village    Italian Restaurant  New American Restaurant   
26  Morningside Heights             Bookstore      American Restaurant   
28    Battery Park City                  Park              Coffee Shop   
29   Financial District           Coffee Shop                Wine Shop   
31                 Noho    Italian Restaurant             Cocktail Bar   
32         Civic Center  Gym / Fitness Center        French Restaurant   
33        Midtown South     Korean Restaurant                    Hotel   
34         Sutton Place  Gym / Fitness Center        Indian Restaurant   
35           Turtle Bay    Italian Restaurant                 Wine Bar   
38             Flatiron  Gym / Fitness Center              Yoga Studio   
39         Hudson Yards   American Restaurant       Italian Restaurant   

   3rd Most Common Venue   4th Most Common Venue   5th Most Common Venue  
1    American Restaurant      Salon / Barbershop                  Bakery  
5     Seafood Restaurant      Mexican Restaurant                    Park  
6    American Restaurant                     Bar          Cosmetics Shop  
8                 Bakery             Art Gallery             Coffee Shop  
10           Pizza Place        Sushi Restaurant            Burger Joint  
11           Coffee Shop   Outdoors & Recreation                  School  
12              Wine Bar             Coffee Shop          Cosmetics Shop  
13  Gym / Fitness Center            Concert Hall      Italian Restaurant  
14   American Restaurant      Italian Restaurant               Wine Shop  
15               Theater          Clothing Store             Coffee Shop  
16   Japanese Restaurant                   Hotel      Italian Restaurant  
17                Bakery      Italian Restaurant               Nightclub  
18        Clothing Store       Indian Restaurant       French Restaurant  
21                  Café                     Spa     American Restaurant  
22                  Café      Salon / Barbershop          Sandwich Place  
23           Art Gallery              Shoe Store  Furniture / Home Store  
24        Cosmetics Shop                    Park     American Restaurant  
26           Coffee Shop                    Park            Burger Joint  
28                 Hotel           Memorial Site                     Gym  
29                 Hotel     American Restaurant             Pizza Place  
31           Art Gallery       French Restaurant      Mexican Restaurant  
32                 Hotel             Coffee Shop      Italian Restaurant  
33             Hotel Bar     Japanese Restaurant          Cosmetics Shop  
34    Italian Restaurant  Furniture / Home Store     American Restaurant  
35            Steakhouse        Sushi Restaurant        Ramen R

Purpose of this project is to recommmend Aya to open her Japanese restaurant in the right place!
As you can see on the result of each cluster, the area in cluster 3 "Stuyvesant Town" is very near from her partner's house.
If she think only where is the most convenience for her to take care the restaurant, she may choose the nearest one!
And she will loss the potential to run the business and she will get loss for sure.

Finally, I would like to suggest her to survey the area that will open the restaurant by herself.
Those area are including 
- Chinatown
- Manhattanville
- Central Harlem
- Upper East Side
- Lenox Hill
- Roosevelt Island
- Upper West Side	
- Lincoln Square
- Clinton
- Midtown
- Murray Hill	
- Chelsea
- Greenwich Village
- Tribeca
- Little Italy
- Soho
- West Village
- Morningside Heights
- Battery Park City
- Financial District
- Noho
- Midtown South
- Civic Center
- Sutton Place
- Turtle Bay
- Flatiron
- Hudson Yards	

Remark, 
This is the result of the area that I limited the distance not over 20 km from the center of Manhattan.